### PR TITLE
Feature/remove statman

### PR DESCRIPTION
**Pull request by ghbutton**

### CODE REVIEW

**Comment by ghbutton** [comment]
Got it to a working state, was not fully able to test since the internet here is limited

File: lib/new_relixir.ex [hunk 1]
```diff
@@ -13,26 +13,12 @@ defmodule NewRelixir do
     import Supervisor.Spec, warn: false
 
     children = [
-      worker(:statman_server, [1000]),
-      worker(:statman_aggregator, []),
+      worker(NewRelixir.Collector, []),
+      worker(NewRelixir.Poller, [&NewRelixir.Statman.poll/0])
```
Comment (ulissesalmeida) [inline lib/new_relixir.ex:17]: I think we need to update the PR description why we're removing the Statman dependency.

File: lib/new_relixir/agent.ex [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  defp app_name() do
+    Application.get_env(:new_relixir, :application_name)
+  end
+
+  defp license_key() do
+    Application.get_env(:new_relixir, :license_key)
+  end
+
+  def request(url, body \\ "[]") do
+    :lhttpc.request(url, :post, [{"Content-Encoding", "identity"}], body, 5000)
```
Comment (ulissesalmeida) [inline lib/new_relixir/agent.ex:116]: We should use this opportunity to get rid off `lhttpc`. It's not supported anymore. Maybe we could use hackney, tesla, or httpoison

> This project is not supported anymore. It's still here just to support legacy projects that may have it as a dependency. It's operational (at least, for Erlang versions lower than R17) If you're looking for a HTTP client, we would recommend you to switch to shotgun or fusco
https://github.com/esl/lhttpc#notice

File: lib/new_relixir/poller.ex [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            round(end_time / 1000),
+            metrics
+          ]
+          try do
+            NewRelixir.Agent.push(hostname, metrics, errors)
+          rescue
+            error -> error_cb.(:push_failed, error)
+          end
+      end
+    rescue
```
Comment (ulissesalmeida) [inline lib/new_relixir/poller.ex:48]: I'm a little bit confused why do we need all these rescues.
If the poller crashes, is not the supervisor responsibility to start again with a clean timer?

File: lib/new_relixir/statman.ex [hunk 4]
```diff
@@ -0,0 +1,186 @@
+defmodule NewRelixir.Statman do
```
Comment (ulissesalmeida) [inline lib/new_relixir/statman.ex:1]: We don't need to call it `Statman` anymore.
I think we can call it `Stats`, or `StatsAggregator` or another better name.

File: lib/new_relixir/statman.ex [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+  def poll do
```
Comment (ulissesalmeida) [inline lib/new_relixir/statman.ex:2]: We could rename it to `pull`

File: lib/new_relixir/collector.ex [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+  def record_value({name, data}, elapsed) do
+    GenServer.cast(@name, {:record_value, {name, data}, elapsed})
+  end
+
+  def record_error(transaction_name, {type, message}) do
+    GenServer.cast(@name, {:record_error, {transaction_name, type, message}})
+  end
+
+  def poll do
```
Comment (ulissesalmeida) [inline lib/new_relixir/collector.ex:18]: I think these ones should be called pull too 🤔

**Review by ulissesalmeida — COMMENTED** [review]
We dropping the `newrelic` dependency. That lib has a public support to Statman that we're removing together. It's a breaking change for other users. It should be a major upgrade in the lib version and I think we should add a CHANGELOG explaining the change and how to upgrade. What do you think?

File: lib/new_relixir.ex [hunk 7]
```diff
@@ -13,26 +13,12 @@ defmodule NewRelixir do
...
-         true        <- String.valid?(app_name),
-         true        <- String.valid?(license_key) do
-      Application.put_env(:newrelic, :application_name, to_char_list(app_name))
-      Application.put_env(:newrelic, :license_key, to_char_list(license_key))
-
-      {:ok, _} = :newrelic_poller.start_link(&:newrelic_statman.poll/0)
-    end
-
-    result
+    Supervisor.start_link(children, opts)
```
Comment (britto) [inline lib/new_relixir.ex:21]: We should consider logging a warning message when the app is missing required configs.

File: lib/new_relixir/poller.ex [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    {:noreply, state}
+  end
+
+  def handle_info(:poll, %{poll_fun: poll_fun, error_cb: error_cb, timer: old_timer}) do
+    :erlang.cancel_timer(old_timer)
+    timer = :erlang.send_after(@poll_interval, self(), :poll)
+    {:ok, hostname} = :inet.gethostname()
+    try do
+      case poll_fun.() do
+        {[], []} ->
```
Comment (britto) [inline lib/new_relixir/poller.ex:34]: I think this rule is never getting matched. Both `webtransaction_total` and `db_total` don't return empty lists, even when there is no data. So we are making "empty" requests to New Relic. We can see it when we log the calls to `push_data`:

```
[debug] Pushing data: [595128166, 1511802221, 1511802251, [[%{name: "HttpDispatcher", scope: ""}, [0, 0, 0, 0, 0, 0]], [%{name: "Database/all", scope: ""}, [0, 0, 0, 0, 0, 0]], [%{name: "Errors/all", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]], [%{name: "Errors/allWeb", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]], [%{name: "Instance/Reporting", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]]]]
[debug] Pushing data: [595128166, []]
[debug] Pushing data: [595129818, 1511802251, 1511802281, [[%{name: "HttpDispatcher", scope: ""}, [0, 0, 0, 0, 0, 0]], [%{name: "Database/all", scope: ""}, [0, 0, 0, 0, 0, 0]], [%{name: "Errors/all", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]], [%{name: "Errors/allWeb", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]], [%{name: "Instance/Reporting", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]]]]
[debug] Pushing data: [595129818, []]
```

File: lib/new_relixir/statman.ex [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    [end_time, start_time, metrics, errors] = NewRelixir.Collector.poll
+    transform_aggregated_metrics(metrics, errors, {start_time, end_time})
+  end
+
+  def transform_aggregated_metrics(metrics, errors, time) do
+    ms = metrics
```
Comment (britto) [inline lib/new_relixir/statman.ex:8]: Since the map itself is enumerable and the output format of `transform_metric` is now standardized, we can simplify this and also skip the filter step:

```elixir
# Before
ms = metrics
      |> Map.to_list
      |> Enum.reduce([], fn(m, acc) ->
        acc ++ transform_metric(m)
      end)
      |> Enum.filter(&(&1 != []))

# After
ms = Enum.flat_map(metrics, &transform_metric/1)
```

File: lib/new_relixir/statman.ex [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+  def errors_total(errors) do
+    data = [length(errors), 0.0, 0.0, 0.0, 0.0, 0.0]
+    [[%{name: "Errors/all", scope: ""}, data],
+     [%{name: "Errors/allWeb", scope: ""}, data],
+     [%{name: "Instance/Reporting", scope: ""}, data]
+    ]
+  end
+
+  def pluck(_, _, []), do: [0]
```
Comment (britto) [inline lib/new_relixir/statman.ex:139]: The logic here is a little hard to follow, I'd recommend refactoring it:

```elixir
# Before
def pluck(_, _, []), do: [0]
def pluck(name, n, list) do
  Enum.map(list, fn(elem) ->
    if is_list(elem) && is_list(hd(elem)) do
      pluck(name, n, elem)
    else
      get_nth(elem, name, n-1)
    end
  end)
  |> List.flatten
end

defp get_nth([[_, []]], _, _), do: 0
defp get_nth([_, []], _, _), do: 0
defp get_nth([[struct, data_list]], name, n), do: get_nth([struct, data_list], name, n)
defp get_nth([struct, data_list], name, n) do
  if String.contains?(struct[:name], name) do
    if struct[:scope] == "" do
      Enum.at(data_list, n) || 0
    else
      0
    end
  else
    0
  end
end

# After
def pluck(type, position, metrics) do
  metrics
  |> Enum.filter(fn [%{name: name}, _data] -> String.contains?(name, type) end)
  |> Enum.map(fn [_metric, data] -> Enum.at(data, position) || 0 end)
end

# Then the calls to `min` and `max` need to provide a default value:

# Before
min  = Enum.min(pluck(name, 4, ms))
max  = Enum.max(pluck(name, 5, ms))

# After
min  = Enum.min(pluck(name, 4, ms), fn -> 0 end)
max  = Enum.max(pluck(name, 5, ms), fn -> 0 end)
```

File: mix.exs [hunk 11]
```diff
@@ -21,20 +20,18 @@ defmodule NewRelixir.Mixfile do
   end
 
   defp deps do
-    [{:phoenix, "~> 1.3-rc"},
-     {:ecto, ">= 1.1.0 and < 3.0.0"},
-     {:newrelic, "~> 0.1.0"},
-     {:earmark, "~> 0.1", only: :dev},
-     {:ex_doc, "~> 0.11", only: :dev}]
+    [
+      {:earmark, "~> 0.1", only: :dev},
```
Comment (britto) [inline mix.exs:24]: The `earmark` library is already a dependency of `ex_doc`, we can skip it here.

**Review by britto — CHANGES_REQUESTED** [review]

File: lib/new_relixir/statman.ex [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    ]
+    List.duplicate(error, count) |> List.flatten
+  end
+
+  defp transform_metric(metric) do
+    transform_histogram(metric)
+  end
+
+
+  def transform_counter(metric) do
```
Comment (britto) [inline lib/new_relixir/statman.ex:40]: This function seems to be unused.

File: lib/new_relixir/statman.ex [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+      [%{
+        parameter_groups: %{},
+        stack_trace: [],
+        request_params: %{},
+        request_uri: scope}]
+    ]
+    List.duplicate(error, count) |> List.flatten
+  end
+
+  defp transform_metric(metric) do
```
Comment (britto) [inline lib/new_relixir/statman.ex:35]: This function doesn't seem to add much value, we should probably call `transform_histogram` straight.

File: lib/new_relixir/poller.ex (hunk 8 above)
Comment (britto) [reply lib/new_relixir/poller.ex:34]: One way we can fix this is adding an extra clause to `transform_aggregated_metrics`, like:

```elixir
def transform_aggregated_metrics([], [], time), do: {[], [], time}
def transform_aggregated_metrics(metrics, errors, time) do
  ms = Enum.flat_map(metrics , &transform_histogram/1)
  errs = Enum.map(errors, &transform_error_counter/1)

  {[webtransaction_total(ms), db_total(ms) | errors_total(errs) ++ ms], errs, time}
end
```

Then match it with the expected arity:

```elixir
case poll_fun.() do
  {[], [], _interval} ->
    Logger.debug "Nothing to send."
    :ok
```

**Comment by ulissesalmeida** [comment]
We can close this one in favor of https://github.com/TheRealReal/new-relixir/pull/24

File: lib/new_relixir/statman.ex (hunk 4 above)
Comment (ulissesalmeida) [reply lib/new_relixir/statman.ex:1]: Done in 03942240549525629e2a5862222f61e82908ea27

File: mix.exs (hunk 11 above)
Comment (ulissesalmeida) [reply mix.exs:24]: Done in 67751f1a3829e7b79319d39be03437c747b90ed2

File: lib/new_relixir/statman.ex (hunk 5 above)
Comment (ulissesalmeida) [reply lib/new_relixir/statman.ex:2]: Done in 74f3035b8a5ea4784aff51333874248fd5ddfc6e

File: lib/new_relixir/collector.ex (hunk 6 above)
Comment (ulissesalmeida) [reply lib/new_relixir/collector.ex:18]: Done in 74f3035b8a5ea4784aff51333874248fd5ddfc6e

File: lib/new_relixir/poller.ex (hunk 8 above)
Comment (ulissesalmeida) [reply lib/new_relixir/poller.ex:34]: Done in 2d4623d2975046a5ff2dc01aa4e564f417e263c2

File: lib/new_relixir/statman.ex (hunk 9 above)
Comment (ulissesalmeida) [reply lib/new_relixir/statman.ex:8]: Done in def81565faac1b97b15e0d9e48a4ad8892c40848

File: lib/new_relixir/statman.ex (hunk 13 above)
Comment (ulissesalmeida) [reply lib/new_relixir/statman.ex:35]: Done in def81565faac1b97b15e0d9e48a4ad8892c40848

File: lib/new_relixir/statman.ex (hunk 12 above)
Comment (ulissesalmeida) [reply lib/new_relixir/statman.ex:40]: Done in 46f98b6ea6f7915ddd1caccf2f8c5d7b437c2cba

File: lib/new_relixir/statman.ex (hunk 10 above)
Comment (ulissesalmeida) [reply lib/new_relixir/statman.ex:139]: Done in 63f7bff23e07f70e761be5202e1cfe58b4092f4d

Comment (britto) [reply lib/new_relixir/statman.ex:139]: Nice! 👌

File: lib/new_relixir/agent.ex (hunk 2 above)
Comment (ulissesalmeida) [reply lib/new_relixir/agent.ex:116]: Done in ea7579301d92225d6b1130a6be5f6dd5126bbc22

File: lib/new_relixir.ex (hunk 7 above)
Comment (ulissesalmeida) [reply lib/new_relixir.ex:21]: Done in 69f55aabd68d9a3e4e4cd3d56ecb1c171aa7af0c
